File: tech_app/backend/services/cost.py
"""
成本分析服务: 对单个零件做专业成本拆解,允许 Claude 联网检索当前行情作为依据。

延续平台思路 —— Claude 只产出**结构化**成本明细(CostAnalysis)并附价格来源,
平台做确定性的金额/合计重算与校验(compute)。
"""
from __future__ import annotations

from typing import List, Optional, Tuple

from ..models.cost import CostAnalysis, WebSource
from ..models.ir import DesignIR, Part
from . import llm_client as claude_client

SYSTEM_PROMPT = """你是资深机械零件成本工程师(应成本/报价)。给定一个零件的结构化设计意图
(特征、材料、尺寸、公差、工艺角色)与可选的几何属性(体积/质量),请做**专业的成本拆解**。

要求:
1. **联网检索当前行情**:用 web_search 工具查当前的原材料价格(如 Q235/45 钢、6061 铝、304
   不锈钢的现货价 元/吨 或 元/kg)、外购标准件价格、表面/热处理与机加工费率等,作为估算依据。
   把检索到的关键价格写进 price_references,**每条尽量给出 source(出处网站) 与 url(可点击的
   具体网页链接) 与 date(价格时间)** —— 让依据可核查、可追溯。分项 items 的 source 字段也尽量
   写明出处。没有检索到链接的价格,请明确标注为"估算"并降低置信度,不要编造网址。
2. 按成本类别拆分 items:材料费、机加工费(工时×费率)、标准件/外购、热处理、表面处理、
   焊接、装配、检验、工装摊销、物流、管理费、利润等(按零件实际涉及的项给出,不要硬凑)。
3. 每个分项尽量给出: 计算依据 basis、数量 quantity、单位 unit、单价 unit_price、金额 amount、
   价格来源 source、置信度 confidence。材料费应基于零件体积/质量×材料密度×单价估算并计损耗。
4. 给出 summary(成本构成与定价思路)、unit_cost(单件成本估算)、assumptions(批量/损耗率/
   费率等关键假设)。凡缺尺寸/材料/批量而影响报价精度的,写进 open_questions 并降低相应置信度。
5. 金额单位统一为人民币元(currency=CNY)。**不要臆造**精确单价 —— 没检索到就标注为估算并降置信度。
6. 全程用中文填写各字段。"""


def _part_prompt(part: Part, overall: Optional[DesignIR], geom: Optional[dict], quantity: int) -> str:
    lines = [f"零件编号: {part.part_id}", f"零件名称: {part.name}", f"核算批量: {quantity} 件"]
    if part.role:
        lines.append(f"装配角色/功能: {part.role}")
    if part.material:
        m = part.material
        lines.append(f"材料: {m.spec}" + (f" (密度 {m.density} g/cm³)" if m.density else ""))
    if part.tolerance_general:
        lines.append(f"一般公差: {part.tolerance_general}")

    if part.features:
        lines.append("特征(几何构成):")
        for i, f in enumerate(part.features, 1):
            dims = ", ".join(
                f"{k}={v}" for k, v in f.model_dump().items()
                if v is not None and k not in ("type", "purpose")
            )
            purpose = f" — {f.purpose}" if f.purpose else ""
            lines.append(f"  {i}. {f.type.value}: {dims}{purpose}")

    if geom:
        g = []
        if geom.get("bbox"):
            g.append("包围盒 " + "×".join(str(x) for x in geom["bbox"]) + " mm")
        if geom.get("volume_mm3"):
            g.append(f"体积 {geom['volume_mm3']} mm³")
        if geom.get("mass_g"):
            g.append(f"质量 {geom['mass_g']} g")
        if g:
            lines.append("几何属性(CAD 内核实测): " + "; ".join(g))

    if overall and overall.device_name:
        lines.append(f"\n所属设备: {overall.device_name}")

    lines.append("\n请先联网检索相关材料/外购/加工的当前行情,再做成本拆解,"
                 "调用工具输出结构化 CostAnalysis。")
    return "\n".join(lines)


def analyze_cost(
    part: Part, overall: Optional[DesignIR] = None, geom: Optional[dict] = None,
    quantity: int = 1, web: bool = True, note: str = "",
    attachments: Optional[List[Tuple[str, bytes]]] = None,
) -> CostAnalysis:
    use_web = web and claude_client.WEB_SEARCH_AVAILABLE
    prompt = _part_prompt(part, overall, geom, quantity)
    content = [claude_client.text_block(prompt), claude_client.text_block(claude_client.web_search_notice(use_web))]
    if note and note.strip():
        content.append(claude_client.text_block(f"【用户补充说明(请优先采用)】\n{note.strip()}"))
    content.extend(claude_client.attachment_blocks(attachments))

    extra_tools = claude_client.web_search_tools(use_web)
    sources: list = []
    analysis = claude_client.run(
        SYSTEM_PROMPT, content, CostAnalysis, extra_tools=extra_tools, sources_out=sources,
    )
    # 确定性归一: 锚定 part_id/name/批量;并把平台收集的检索来源合并进 search_sources(去重)
    analysis.part_id = part.part_id
    if not analysis.part_name:
        analysis.part_name = part.name
    if not analysis.material and part.material:
        analysis.material = part.material.spec
    analysis.quantity = quantity

    have = {s.url for s in analysis.search_sources}
    for s in sources:
        if s.get("url") and s["url"] not in have:
            analysis.search_sources.append(WebSource(title=s.get("title") or s["url"], url=s["url"]))
            have.add(s["url"])
    return analysis


def compute(analysis: dict) -> dict:
    """确定性派生量: 按明细重算金额与合计、分类汇总、勾稽校验告警。"""
    items = analysis.get("items", []) or []
    warnings: list[str] = []
    total = 0.0
    by_category: dict[str, float] = {}

    for it in items:
        amt = it.get("amount")
        qty = it.get("quantity")
        up = it.get("unit_price")
        # 数量×单价 与 金额 勾稽
        if isinstance(qty, (int, float)) and isinstance(up, (int, float)):
            calc = round(qty * up, 2)
            if amt is None:
                amt = calc
            elif abs(calc - amt) > max(0.01, abs(amt) * 0.02):
                warnings.append(
                    f"分项「{it.get('name')}」金额 {amt} 与 数量×单价 ({calc}) 不符")
        if isinstance(amt, (int, float)):
            total += amt
            cat = it.get("category", "other")
            by_category[cat] = round(by_category.get(cat, 0.0) + amt, 2)

    computed_total = round(total, 2) if items else None
    stated = analysis.get("unit_cost")
    if isinstance(stated, (int, float)) and computed_total is not None:
        if abs(stated - computed_total) > max(0.5, abs(computed_total) * 0.05):
            warnings.append(f"声明单件成本 {stated} 与 明细合计 {computed_total} 不一致")

    return {
        "item_count": len(items),
        "computed_total": computed_total,
        "by_category": by_category,
        "currency": analysis.get("currency", "CNY"),
        "warnings": warnings,
    }

File: tech_app/frontend/cost.js
// 成本分析页: 读取 ?project=&part=, 生成/展示/编辑该零件的结构化成本拆解(支持联网检索行情)。
const API = "";
const authToken = localStorage.getItem("authToken") || "";

const _fetch = window.fetch.bind(window);
window.fetch = (url, opts = {}) => {
  const hasAuth = opts.headers && (opts.headers.Authorization || opts.headers.authorization);
  if (typeof url === "string" && url.indexOf("/api/") !== -1 && authToken && !hasAuth) {
    opts = Object.assign({}, opts, {
      headers: Object.assign({}, opts.headers, { Authorization: "Bearer " + authToken }),
    });
  }
  return _fetch(url, opts).then(r => { if (r.status === 401) location.href = "index.html"; return r; });
};

const $ = (id) => document.getElementById(id);
const status = (m, busy = false) => { $("status").textContent = (busy ? "⏳ " : "") + m; };
const esc = (s) => String(s ?? "").replace(/[&<>]/g, c => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;" }[c]));
const sleep = (ms) => new Promise(r => setTimeout(r, ms));
const money = (v) => (v == null ? "—" : Number(v).toLocaleString("zh-CN", { maximumFractionDigits: 2 }));

const CAT_LABEL = {
  material: "材料费", machining: "机加工", standard_part: "标准件/外购", heat_treat: "热处理",
  surface: "表面处理", welding: "焊接", assembly: "装配", inspection: "检验",
  tooling: "工装摊销", logistics: "物流包装", overhead: "管理费", profit: "利润", other: "其他",
};
const CATS = Object.keys(CAT_LABEL);

const qs = new URLSearchParams(location.search);
const PID = qs.get("project");
const PART = qs.get("part");

let part = null, analysis = null, summary = null, editing = false;

$("extraFiles").addEventListener("change", () => {
  const names = [...$("extraFiles").files].map(file => file.name);
  $("extraFilesName").textContent = names.length ? names.join("、") : "未选择文件";
});

async function init() {
  try { const h = await fetch(`${API}/api/health`).then(r => r.json()); $("health").textContent = ""; }
  catch { $("health").textContent = "后端未连接"; }

  if (!PID || !PART) { status("缺少 project / part 参数,请从工作台点击零件进入。"); return; }
  $("backLink").href = `index.html?project=${encodeURIComponent(PID)}&part=${encodeURIComponent(PART)}`;

  try {
    const data = await fetch(`${API}/api/projects/${PID}`).then(r => r.json());
    part = (data.ir && data.ir.parts || []).find(p => p.part_id === PART);
  } catch { /* ignore */ }
  if (!part) { $("partName").textContent = "(未找到该零件)"; status("未找到零件,请确认已解析。"); return; }
  $("partId").textContent = part.part_id;
  $("partName").textContent = part.name || part.part_id;
  $("partMat").textContent = (part.material ? "材料 " + part.material.spec : "");

  await loadCost();
}

async function loadCost() {
  try {
    const d = await fetch(`${API}/api/projects/${PID}/parts/${PART}/cost`).then(r => r.json());
    analysis = d.analysis; summary = d.summary;
  } catch { analysis = null; }
  if (analysis) {
    $("btnGen").textContent = "🔄 重新生成";
    $("btnEdit").disabled = false;
    if (analysis.quantity) $("qtyInput").value = analysis.quantity;
    render();
    status("已加载成本分析。可「编辑」改价后保存,或「重新生成」。");
  } else {
    $("costArea").innerHTML = `<div class="panel-card"><div class="row">尚未生成成本分析。` +
      `设定批量后点击「💰 生成成本分析」,AI 将<b>联网检索当前材料/外购/加工行情</b>,` +
      `并按材料/加工/表面/管理/利润等拆解结构化成本。</div></div>`;
    status("尚未生成成本分析。");
  }
}

$("btnGen").onclick = async () => {
  const qty = Math.max(1, parseInt($("qtyInput").value) || 1);
  $("btnGen").disabled = true;
  status("已提交成本分析任务，AI 正在拆解成本；仅在当前模型支持时检索公开行情（较耗时）…", true);
  try {
    const fd = new FormData();
    fd.append("note", $("extraNote").value || "");
    for (const f of $("extraFiles").files) fd.append("attachments", f);
    const sub = await fetch(`${API}/api/projects/${PID}/parts/${PART}/cost?quantity=${qty}`,
      { method: "POST", body: fd });
    const sd = await sub.json();
    if (!sub.ok) throw new Error(sd.detail || sub.status);
    const res = await pollTask(sd.task_id);
    analysis = res.analysis; summary = res.summary; editing = false;
    $("btnGen").textContent = "🔄 重新生成"; $("btnEdit").disabled = false;
    render();
    status(`成本分析完成:${summary.item_count} 个分项,单件约 ${money(summary.computed_total)} 元`);
  } catch (e) { status("成本分析失败: " + e.message); }
  finally { $("btnGen").disabled = false; }
};

async function pollTask(taskId) {
  while (true) {
    await sleep(1500);
    let t;
    try { t = await fetch(`${API}/api/projects/${PID}/tasks/${taskId}`).then(r => r.json()); }
    catch { continue; }
    if (t.status === "succeeded") return t.result;
    if (t.status === "failed") throw new Error(t.error || "任务失败");
    status(`成本分析：${t.progress || "正在处理"}…`, true);
  }
}

$("btnEdit").onclick = () => { editing = true; $("btnEdit").style.display = "none"; $("btnSave").style.display = ""; render(); };

$("btnSave").onclick = async () => {
  collectEdits();
  status("保存成本分析…", true);
  try {
    const r = await fetch(`${API}/api/projects/${PID}/parts/${PART}/cost`, {
      method: "PUT", headers: { "Content-Type": "application/json" }, body: JSON.stringify(analysis),
    });
    const d = await r.json();
    if (!r.ok) throw new Error(d.detail || r.status);
    analysis = d.analysis; summary = d.summary;
    editing = false; $("btnSave").style.display = "none"; $("btnEdit").style.display = "";
    render(); status("已保存(金额与合计已重算)。");
  } catch (e) { status("保存失败: " + e.message); }
};

function render() {
  if (!analysis) return;
  const cur = (summary && summary.currency) || analysis.currency || "CNY";
  let h = "";

  // 概览
  h += `<section class="panel-card"><div class="card-h">成本概览</div>`;
  h += `<div class="cost-total"><span class="v">${money(summary && summary.computed_total)}</span>` +
    `<span class="u">元 / 件 (${esc(cur)})</span><span class="q">核算批量 ${analysis.quantity || 1} 件</span></div>`;
  if (analysis.summary) h += `<div class="row">${esc(analysis.summary)}</div>`;
  // 分类汇总条
  const bc = (summary && summary.by_category) || {};
  const maxv = Math.max(1, ...Object.values(bc));
  const cats = Object.keys(bc).sort((a, b) => bc[b] - bc[a]);
  if (cats.length) {
    h += `<div class="cat-bars">`;
    cats.forEach(cat => {
      const v = bc[cat];
      h += `<div class="cat-bar"><div class="cb-top"><span>${CAT_LABEL[cat] || cat}</span><span>${money(v)} 元</span></div>` +
        `<div class="cb-track"><div class="cb-fill" style="width:${(v / maxv * 100) | 0}%"></div></div></div>`;
    });
    h += `</div>`;
  }
  (summary && summary.warnings || []).forEach(w => h += `<div class="proc-warn">⚠ ${esc(w)}</div>`);
  h += `</section>`;

  // 成本明细表
  h += `<section class="panel-card"><div class="card-h">成本明细</div><table class="cost-table"><thead><tr>` +
    `<th>类别</th><th>分项</th><th>计算依据</th><th>数量</th><th>单位</th><th>单价</th><th>金额(元)</th><th>来源</th><th>置信</th>` +
    `</tr></thead><tbody>`;
  (analysis.items || []).forEach((it, i) => h += itemRow(it, i));
  h += `</tbody></table>`;
  if (editing) h += `<div class="row edit-hint">提示：保存后平台会按 数量×单价 重算金额与合计。</div>`;
  h += `</section>`;

  // 价格依据(联网检索,带可点击链接)
  if ((analysis.price_references || []).length) {
    h += `<section class="panel-card"><div class="card-h">价格依据(联网检索)</div>`;
    analysis.price_references.forEach(r => {
      const link = r.url ? ` · <a href="${esc(r.url)}" target="_blank" rel="noopener">查看来源 ↗</a>` : "";
      h += `<div class="priceref"><span class="pi">${esc(r.item)}</span> — <span class="pp">${esc(r.price)}</span>` +
        `<div class="ps">${esc(r.source || "")}${r.date ? " · " + esc(r.date) : ""}${link}</div></div>`;
    });
    h += `</section>`;
  }

  // 检索来源(平台自动收集的网页证据)
  if ((analysis.search_sources || []).length) {
    h += `<section class="panel-card"><div class="card-h">检索来源(可点击核查)</div>`;
    analysis.search_sources.forEach(s => {
      h += `<div class="srclink">🔗 <a href="${esc(s.url)}" target="_blank" rel="noopener">${esc(s.title || s.url)}</a></div>`;
    });
    h += `</section>`;
  }

  // 假设 + 待澄清
  if ((analysis.assumptions || []).length || (analysis.open_questions || []).length) {
    h += `<section class="panel-card"><div class="card-h">估算假设 / 待澄清</div>`;
    (analysis.assumptions || []).forEach(a => h += `<div class="assump">• ${esc(a)}</div>`);
    (analysis.open_questions || []).forEach(q =>
      h += `<div class="proc-q">❓ ${esc(q.field)}: ${esc(q.reason)}${q.guess ? " (猜测: " + esc(q.guess) + ")" : ""}</div>`);
    h += `</section>`;
  }

  $("costArea").innerHTML = h;
}

function itemRow(it, i) {
  const conf = ((it.confidence || 0) * 100 | 0) + "%";
  if (editing) {
    const opts = CATS.map(c => `<option value="${c}" ${c === it.category ? "selected" : ""}>${CAT_LABEL[c]}</option>`).join("");
    return `<tr data-i="${i}">` +
      `<td><select data-f="category" style="font-size:12px">${opts}</select></td>` +
      `<td><input data-f="name" value="${esc(it.name || "")}"/></td>` +
      `<td><input data-f="basis" value="${esc(it.basis || "")}"/></td>` +
      `<td><input class="num" data-f="quantity" type="number" step="any" value="${it.quantity != null ? it.quantity : ""}"/></td>` +
      `<td><input data-f="unit" value="${esc(it.unit || "")}" style="width:48px"/></td>` +
      `<td><input class="num" data-f="unit_price" type="number" step="any" value="${it.unit_price != null ? it.unit_price : ""}"/></td>` +
      `<td><input class="num" data-f="amount" type="number" step="any" value="${it.amount != null ? it.amount : ""}"/></td>` +
      `<td><input data-f="source" value="${esc(it.source || "")}"/></td>` +
      `<td>${conf}</td></tr>`;
  }
  return `<tr>` +
    `<td><span class="cat-tag">${CAT_LABEL[it.category] || it.category}</span></td>` +
    `<td>${esc(it.name)}</td><td class="src">${esc(it.basis || "")}</td>` +
    `<td class="num">${it.quantity != null ? it.quantity : ""}</td><td>${esc(it.unit || "")}</td>` +
    `<td class="num">${it.unit_price != null ? money(it.unit_price) : ""}</td>` +
    `<td class="num amt">${it.amount != null ? money(it.amount) : ""}</td>` +
    `<td class="src">${esc(it.source || "")}</td><td class="num">${conf}</td></tr>`;
}

function collectEdits() {
  document.querySelectorAll(".cost-table tr[data-i]").forEach(el => {
    const it = analysis.items[+el.dataset.i];
    el.querySelectorAll("[data-f]").forEach(inp => {
      const f = inp.dataset.f, v = inp.value;
      if (["quantity", "unit_price", "amount"].includes(f)) it[f] = v.trim() === "" ? null : parseFloat(v);
      else it[f] = v.trim() === "" ? (f === "category" ? "other" : null) : v;
    });
  });
}

init();
